File: PHP/src/Model/Shipment.php
<?php

declare(strict_types=1);

namespace Kargonomi\Model;

final readonly class Shipment
{
    /**
     * @param list<ShipmentPackage> $packages
     * @param array<string, mixed> $extra
     */
    public function __construct(
        public int $id,
        public ShipmentStatus $status,
        public ?string $statusLabel,
        public ?string $type,
        public ?int $packageCount,
        public ?string $estimatedPrice,
        public ?string $realPrice,
        public ?string $extraShippingPrice,
        public ?string $buyerName,
        public ?string $shippingWebserviceOrderId,
        public ?string $shippingWebserviceBarcode,
        public ?string $shippingWebserviceTrackingCode,
        public ?string $shippingProviderName,
        public ?string $shippingProviderSlug,
        public ?string $barcodeOfOrderId,
        public ?\DateTimeImmutable $shippingWebserviceCreatedAt,
        public ?string $ecommerceProviderOrderNo,
        public ?string $ecommerceProvider,
        public ?\DateTimeImmutable $deliveryDateToShipmentOffice,
        public ?\DateTimeImmutable $shippingProviderCustomerDeliveryDate,
        public ?\DateTimeImmutable $createdAt,
        public ?\DateTimeImmutable $updatedAt,
        public ?ShipmentParty $sender,
        public ?ShipmentParty $buyer,
        public ?ShipmentPricing $pricing,
        public ?ShipmentWarehouse $warehouse,
        public array $packages,
        public array $extra,
    ) {}

    /** @param array<string, mixed> $data */
    public static function fromArray(array $data): self
    {
        $known = array_fill_keys([
            'id', 'status', 'status_label', 'type', 'package_count', 'estimated_price', 'real_price',
            'extra_shipping_price', 'buyer_name', 'created_at', 'updated_at', 'shipment_packages',
            'shipping_webservice_order_id', 'shipping_webservice_barcode', 'shipping_webservice_tracking_code',
            'shipping_provider_name', 'shipping_provider_slug', 'barcode_of_order_id', 'shipping_webservice_created_at',
            'ecommerce_provider_order_no', 'ecommerce_provider', 'delivery_date_to_shipment_office',
            'shipping_provider_customer_delivery_date', 'sender', 'buyer', 'pricing', 'warehouse',
        ], true);
        $sender = self::object($data['sender'] ?? null);
        $buyer = self::object($data['buyer'] ?? null);
        $pricing = self::object($data['pricing'] ?? null);
        $warehouse = self::object($data['warehouse'] ?? null);

        return new self(
            self::integer($data, 'id'),
            new ShipmentStatus(self::text($data, 'status')),
            self::nullableText($data['status_label'] ?? null),
            self::nullableText($data['type'] ?? null),
            self::nullableInteger($data['package_count'] ?? null),
            self::decimal($data['estimated_price'] ?? null),
            self::decimal($data['real_price'] ?? null),
            self::decimal($data['extra_shipping_price'] ?? null),
            self::nullableText($data['buyer_name'] ?? null),
            self::nullableText($data['shipping_webservice_order_id'] ?? null),
            self::nullableText($data['shipping_webservice_barcode'] ?? null),
            self::nullableText($data['shipping_webservice_tracking_code'] ?? null),
            self::nullableText($data['shipping_provider_name'] ?? null),
            self::nullableText($data['shipping_provider_slug'] ?? null),
            self::nullableText($data['barcode_of_order_id'] ?? null),
            self::date($data['shipping_webservice_created_at'] ?? null),
            self::nullableText($data['ecommerce_provider_order_no'] ?? null),
            self::nullableText($data['ecommerce_provider'] ?? null),
            self::date($data['delivery_date_to_shipment_office'] ?? null),
            self::date($data['shipping_provider_customer_delivery_date'] ?? null),
            self::date($data['created_at'] ?? null),
            self::date($data['updated_at'] ?? null),
            $sender === null ? null : ShipmentParty::fromArray($sender, 'sender'),
            $buyer === null ? null : ShipmentParty::fromArray($buyer, 'buyer'),
            $pricing === null ? null : ShipmentPricing::fromArray($pricing),
            $warehouse === null ? null : ShipmentWarehouse::fromArray($warehouse),
            self::packages($data['shipment_packages'] ?? []),
            array_diff_key($data, $known),
        );
    }

    /** @param array<string, mixed> $data */
    private static function integer(array $data, string $key): int
    {
        if (!isset($data[$key]) || !is_numeric($data[$key])) {
            throw new \UnexpectedValueException("Shipment {$key} must be numeric.");
        }

        return (int) $data[$key];
    }

    /** @param array<string, mixed> $data */
    private static function text(array $data, string $key): string
    {
        if (!isset($data[$key]) || !is_string($data[$key])) {
            throw new \UnexpectedValueException("Shipment {$key} must be a string.");
        }

        return $data[$key];
    }

    private static function nullableText(mixed $value): ?string
    {
        return is_string($value) ? $value : null;
    }

    private static function decimal(mixed $value): ?string
    {
        if ($value === null) {
            return null;
        }
        if (!is_string($value) && !is_int($value) && !is_float($value)) {
            throw new \UnexpectedValueException('Price must be represented by a string or number.');
        }

        return (string) $value;
    }

    private static function nullableInteger(mixed $value): ?int
    {
        if ($value === null) {
            return null;
        }
        if (!is_int($value) && !(is_string($value) && ctype_digit($value))) {
            throw new \UnexpectedValueException('Package count must be an integer.');
        }

        return (int) $value;
    }

    private static function date(mixed $value): ?\DateTimeImmutable
    {
        if (!is_string($value) || $value === '') {
            return null;
        }

        return new \DateTimeImmutable($value);
    }

    /** @return list<ShipmentPackage> */
    private static function packages(mixed $value): array
    {
        if (!is_array($value)) {
            return [];
        }

        $result = [];
        foreach ($value as $item) {
            $normalized = self::object($item);
            if ($normalized === null) {
                continue;
            }
            $result[] = ShipmentPackage::fromArray($normalized);
        }

        return $result;
    }

    /** @return array<string, mixed>|null */
    private static function object(mixed $value): ?array
    {
        if (!is_array($value) || array_is_list($value)) {
            return null;
        }
        $result = [];
        foreach ($value as $key => $field) {
            if (is_string($key)) {
                $result[$key] = $field;
            }
        }

        return $result;
    }
}

File: ASP.net/KargonomiAsp/Models/Shipment.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Kargonomi.Client.Models;

public sealed record Shipment
{
    [JsonPropertyName("id")]
    public long Id { get; init; }
    [JsonPropertyName("type")]
    public string Type { get; init; } = string.Empty;
    [JsonPropertyName("shipping_webservice_order_id")]
    public string? ShippingWebserviceOrderId { get; init; }
    [JsonPropertyName("shipping_webservice_barcode")]
    public string? ShippingWebserviceBarcode { get; init; }
    [JsonPropertyName("shipping_webservice_tracking_code")]
    public string? ShippingWebserviceTrackingCode { get; init; }
    [JsonPropertyName("shipping_provider_name")]
    public string? ShippingProviderName { get; init; }
    [JsonPropertyName("shipping_provider_slug")]
    public string? ShippingProviderSlug { get; init; }
    [JsonPropertyName("barcode_of_order_id")]
    public string? BarcodeOfOrderId { get; init; }
    [JsonPropertyName("status")]
    public ShipmentStatus Status { get; init; }
    [JsonPropertyName("status_label")]
    public string StatusLabel { get; init; } = string.Empty;
    [JsonPropertyName("shipping_webservice_created_at")]
    public DateTimeOffset? ShippingWebserviceCreatedAt { get; init; }
    [JsonPropertyName("ecommerce_provider_order_no")]
    public string? EcommerceProviderOrderNo { get; init; }
    [JsonPropertyName("ecommerce_provider")]
    public string? EcommerceProvider { get; init; }
    [JsonPropertyName("package_count")]
    public int PackageCount { get; init; }
    [JsonPropertyName("estimated_price")]
    public JsonElement? EstimatedPrice { get; init; }
    [JsonPropertyName("real_price")]
    public JsonElement? RealPrice { get; init; }
    [JsonPropertyName("extra_shipping_price")]
    public string? ExtraShippingPrice { get; init; }
    [JsonPropertyName("buyer_name")]
    public string? BuyerName { get; init; }
    [JsonPropertyName("delivery_date_to_shipment_office")]
    public DateTimeOffset? DeliveryDateToShipmentOffice { get; init; }
    [JsonPropertyName("shipping_provider_customer_delivery_date")]
    public DateTimeOffset? ShippingProviderCustomerDeliveryDate { get; init; }
    [JsonPropertyName("created_at")]
    public DateTimeOffset CreatedAt { get; init; }
    [JsonPropertyName("updated_at")]
    public DateTimeOffset UpdatedAt { get; init; }
    [JsonPropertyName("sender")]
    public ShipmentParty? Sender { get; init; }
    [JsonPropertyName("buyer")]
    public ShipmentParty? Buyer { get; init; }
    [JsonPropertyName("pricing")]
    public ShipmentPricing? Pricing { get; init; }
    [JsonPropertyName("warehouse")]
    public Warehouse? Warehouse { get; init; }
    [JsonPropertyName("shipment_packages")]
    public IReadOnlyList<ShipmentPackage> Packages { get; init; } = [];
    [JsonExtensionData]
    public IDictionary<string, JsonElement> AdditionalData { get; init; } = new Dictionary<string, JsonElement>();
}
public sealed record ShipmentParty(
    [property: JsonPropertyName("sender_name")] string? SenderName,
    [property: JsonPropertyName("buyer_name")] string? BuyerName,
    [property: JsonPropertyName("sender_email")] string? SenderEmail,
    [property: JsonPropertyName("buyer_email")] string? BuyerEmail,
    [property: JsonPropertyName("sender_phone")] string? SenderPhone,
    [property: JsonPropertyName("buyer_phone")] string? BuyerPhone,
    [property: JsonPropertyName("sender_phone2")] string? SenderPhone2,
    [property: JsonPropertyName("buyer_phone2")] string? BuyerPhone2,
    [property: JsonPropertyName("sender_tax_number")] string? SenderTaxNumber,
    [property: JsonPropertyName("buyer_tax_number")] string? BuyerTaxNumber,
    [property: JsonPropertyName("sender_tax_place")] string? SenderTaxPlace,
    [property: JsonPropertyName("buyer_tax_place")] string? BuyerTaxPlace,
    [property: JsonPropertyName("sender_address")] string? SenderAddress,
    [property: JsonPropertyName("buyer_address")] string? BuyerAddress,
    [property: JsonPropertyName("sender_state")] string? SenderState,
    [property: JsonPropertyName("buyer_state")] string? BuyerState,
    [property: JsonPropertyName("sender_city")] string? SenderCity,
    [property: JsonPropertyName("buyer_city")] string? BuyerCity);
public sealed record ShipmentPricing(
    [property: JsonPropertyName("package_count")] int PackageCount,
    [property: JsonPropertyName("estimated_price")] JsonElement? EstimatedPrice,
    [property: JsonPropertyName("real_price")] JsonElement? RealPrice,
    [property: JsonPropertyName("extra_shipping_price")] string? ExtraShippingPrice,
    [property: JsonPropertyName("price_diff")] JsonElement? PriceDifference);
public sealed record Warehouse
{
    [JsonPropertyName("id")]
    public long Id { get; init; }
    [JsonPropertyName("name")]
    public string Name { get; init; } = string.Empty;
    [JsonPropertyName("is_main")]
    public JsonElement IsMain { get; init; }
    [JsonPropertyName("contact_name")]
    public string? ContactName { get; init; }
    [JsonPropertyName("contact_phone")]
    public string? ContactPhone { get; init; }
    [JsonPropertyName("state")]
    public string? State { get; init; }
    [JsonPropertyName("city")]
    public string? City { get; init; }
    [JsonPropertyName("address")]
    public string? Address { get; init; }
    [JsonPropertyName("phone")]
    public string? Phone { get; init; }
    [JsonPropertyName("tax_number")]
    public string? TaxNumber { get; init; }
    [JsonPropertyName("tax_place")]
    public string? TaxPlace { get; init; }
}
public sealed record ShipmentPackage(
    [property: JsonPropertyName("desi")] string Desi,
    [property: JsonPropertyName("barcode")] string? Barcode,
    [property: JsonPropertyName("content")] string? Content,
    [property: JsonPropertyName("real_desi")] string? RealDesi);
